include/fusilli/node/matmul_node.h

```diff
@@ -0,0 +1,204 @@
+    int64_t bK = bDim[bRank - 2]; // Second-to-last dimension of B
+
+    FUSILLI_RETURN_ERROR_IF(
+        aK != bK, ErrorCode::InvalidAttribute,
+        "Matmul input tensors A and B have incompatible inner dimensions (K): "
+        "A has K=" +
+            std::to_string(aK) + ", B has K=" + std::to_string(bK));
+
+    // Check that batch dimensions are broadcastable.
+    // Since both inputs have the same rank, we can directly compare batch dims.
+    size_t batchDims = aRank - kNonBatchRank;
+    for (size_t i = 0; i < batchDims; ++i) {
+      int64_t aDimVal = aDim[i];
+      int64_t bDimVal = bDim[i];
+      FUSILLI_RETURN_ERROR_IF(
+          !(aDimVal % bDimVal == 0 || bDimVal % aDimVal == 0),
+          ErrorCode::InvalidAttribute,
+          "Matmul input tensors A and B have incompatible batch dimensions for "
+          "broadcasting at index " +
+              std::to_string(i) + ": A has dim=" + std::to_string(aDimVal) +
+              ", B has dim=" + std::to_string(bDimVal));
+    }
+
+    return ok();
+  }
+
+  ErrorObject inferPropertiesNode() override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Inferring properties for MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    matmulAttr.fillFromContext(context);
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    const std::vector<int64_t> &aDim = aT->getDim();
+    const std::vector<int64_t> &bDim = bT->getDim();
+
+    const std::vector<int64_t> &cDim = cT->getDim();
+    const std::vector<int64_t> &cStride = cT->getStride();
+
+    // Infer shape of output tensor.
+    if (cDim.empty())
+      cT->setDim(getMatmulInferredOutputShape(aDim, bDim));
+
+    // Output stride is contiguous (row-major) when unspecified.
+    if (cStride.empty()) {
+      cT->setStride(
+          generateStrideFromDim(cDim, getContiguousStrideOrder(cDim.size())));
+    }
+
+    return ok();
+  }
+
+  ErrorObject postValidateNode() const override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Post-Validating MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    size_t cRank = cT->getDim().size();
+
+    // Rank checks
+    constexpr int64_t kNonBatchRank = 2;
+    FUSILLI_RETURN_ERROR_IF(
+        cRank < kNonBatchRank, ErrorCode::InvalidAttribute,
+        "Matmul output tensor C must have a rank of at least 2");
+
+    FUSILLI_RETURN_ERROR_IF(
+        cT->getDim() !=
+            getMatmulInferredOutputShape(aT->getDim(), bT->getDim()),
+        ErrorCode::InvalidAttribute,
+        "Matmul output tensor C dimensions do not match the expected shapes "
+        "inferred based on the input dimensions");
+    return ok();
```

Comment:
As far as I understand, there might be situation when user manually sets `cDim` and `cStride` (they are not inferred by `inferPropertiesNode`).

Probably we should validate `cStride` here too? For example, need to check that `cStride` has the rank `cRank`.